sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/VertxAsyncHttpClient.java

```diff
@@ -0,0 +1,115 @@
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+/**
+ * {@link HttpClient} implementation for the Vert.x {@link WebClient}.
+ */
+public class VertxAsyncHttpClient implements HttpClient {
+
+    private static final ClientLogger LOGGER = new ClientLogger(VertxAsyncHttpClient.class);
+    final WebClient client;
+
+    /**
+     * Constructs a {@link VertxAsyncHttpClient}.
+     *
+     * @param client The Vert.x {@link WebClient}
+     */
+    public VertxAsyncHttpClient(WebClient client) {
+        Objects.requireNonNull(client, "client cannot be null");
+        this.client = client;
+    }
+
+    @Override
+    public Mono<HttpResponse> send(HttpRequest request) {
+        return send(request, Context.NONE);
+    }
+
+    @Override
+    public Mono<HttpResponse> send(HttpRequest request, Context context) {
+        boolean eagerlyReadResponse = (boolean) context.getData("azure-eagerly-read-response").orElse(false);
+        return Mono.create(sink -> sink.onRequest(value -> {
+            toVertxHttpRequest(request).subscribe(vertxHttpRequest -> {
+                vertxHttpRequest.send(new VertxHttpResponseHandler(request, sink, eagerlyReadResponse));
+            }, sink::error);
+        }));
+    }
+
+    String getRequestUrlAsString(HttpRequest request) {
+        URL url = request.getUrl();
+        if (url.getPath().isEmpty()) {
+            try {
+                url = new URL(url.getProtocol(), url.getHost(), url.getPort(), "/" + url.getFile());
+            } catch (MalformedURLException e) {
+                throw LOGGER.logThrowableAsError(new RuntimeException(e));
+            }
+        }
+        return url.toString();
+    }
+
+
+    private Mono<VertxHttpRequest> toVertxHttpRequest(HttpRequest request) {
+        return Mono.from(convertBodyToBuffer(request))
+                .map(buffer -> {
+                    HttpMethod httpMethod = request.getHttpMethod();
+                    io.vertx.core.http.HttpMethod requestMethod = io.vertx.core.http.HttpMethod.valueOf(httpMethod.name());
+
+                    io.vertx.ext.web.client.HttpRequest<Buffer> delegate = client
+                            .requestAbs(requestMethod, getRequestUrlAsString(request));
+
+                    if (request.getHeaders() != null) {
+                        request.getHeaders()
+                                .stream()
+                                .forEach(httpHeader -> delegate.putHeader(httpHeader.getName(),
+                                        httpHeader.getValuesList()));
+                    }
+
+                    return new VertxHttpRequest(delegate, buffer);
+                });
+    }
+
+    private Mono<Buffer> convertBodyToBuffer(HttpRequest request) {
+        return Mono.using(() -> Buffer.buffer(),
+                buffer -> getBody(request).reduce(buffer, (b, byteBuffer) -> {
+                    for (int i = 0; i < byteBuffer.limit(); i++) {
+                        b.appendByte(byteBuffer.get(i));
```

Comment:
Yes (memory permitting).

Reply:
@jamesnetherton Have you explored possibility to use https://vertx.io/docs/apidocs/io/vertx/ext/web/client/HttpRequest.html#sendStream-io.vertx.core.streams.ReadStream-io.vertx.core.Handler- to avoid buffering ?

Reply:
Only briefly, but I never arrived at a solution that worked so I abandoned work on it.

I could try to take another look into it.

Reply:
Latest commit switches from the 'high level' `WebClient` to the 'low level' Vert.x `HttpClient` for finer control over request / response bodies. They're now streamed to avoid buffering.